internal/pkg/provision/providers/firecracker/node.go

```diff
@@ -0,0 +1,195 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package firecracker
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"net"
+	"os"
+	"path/filepath"
+
+	"github.com/firecracker-microvm/firecracker-go-sdk"
+	models "github.com/firecracker-microvm/firecracker-go-sdk/client/models"
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/talos-systems/talos/internal/pkg/kernel"
+	"github.com/talos-systems/talos/internal/pkg/provision"
+)
+
+func (p *provisioner) createDisk(state *state, nodeReq provision.NodeRequest) (diskPath string, err error) {
+	diskPath = filepath.Join(state.tempDir, fmt.Sprintf("%s.disk", nodeReq.Name))
+
+	var diskF *os.File
+
+	diskF, err = os.Create(diskPath)
+	if err != nil {
+		return
+	}
+
+	defer diskF.Close() //nolint: errcheck
+
+	err = diskF.Truncate(nodeReq.DiskSize)
+
+	return
+}
+
+func (p *provisioner) createNodes(ctx context.Context, state *state, clusterReq provision.ClusterRequest, nodeReqs []provision.NodeRequest) ([]provision.NodeInfo, error) {
+	errCh := make(chan error)
+	nodeCh := make(chan provision.NodeInfo, len(nodeReqs))
+
+	for _, nodeReq := range nodeReqs {
+		go func(nodeReq provision.NodeRequest) {
+			nodeInfo, err := p.createNode(ctx, state, clusterReq, nodeReq)
+			errCh <- err
+
+			if err == nil {
+				nodeCh <- nodeInfo
+			}
+		}(nodeReq)
+	}
+
+	var multiErr *multierror.Error
+
+	for range nodeReqs {
+		multiErr = multierror.Append(multiErr, <-errCh)
+	}
+
+	close(nodeCh)
+
+	nodesInfo := make([]provision.NodeInfo, 0, len(nodeReqs))
+
+	for nodeInfo := range nodeCh {
+		nodesInfo = append(nodesInfo, nodeInfo)
+	}
+
+	return nodesInfo, multiErr.ErrorOrNil()
+}
+
+func (p *provisioner) createNode(ctx context.Context, state *state, clusterReq provision.ClusterRequest, nodeReq provision.NodeRequest) (provision.NodeInfo, error) {
+	socketPath := filepath.Join(state.tempDir, fmt.Sprintf("%s.sock", nodeReq.Name))
+
+	vcpuCount := int64(math.RoundToEven(float64(nodeReq.NanoCPUs) / 1000 / 1000 / 1000))
+	if vcpuCount < 2 {
+		vcpuCount = 1
+	}
+
+	memSize := nodeReq.Memory / 1024 / 1024
+
+	diskPath, err := p.createDisk(state, nodeReq)
+	if err != nil {
+		return provision.NodeInfo{}, err
+	}
+
+	cmdline := kernel.NewDefaultCmdline()
+
+	// required to get kernel console
+	cmdline.Append("console", "ttyS0")
+
+	// reboot configuration
+	cmdline.Append("reboot", "k")
+	cmdline.Append("panic", "1")
+
+	// disable stuff we don't need
+	cmdline.Append("pci", "off")
+	cmdline.Append("acpi", "off")
+	cmdline.Append("i8042.noaux", "")
+
+	// Talos config
+	cmdline.Append("talos.platform", "metal")
+	cmdline.Append("talos.config", fmt.Sprintf("%s%s.yaml", state.baseConfigURL, nodeReq.Name))
+
+	// networking
+	cmdline.Append("ip", fmt.Sprintf(
+		"%s::%s:%s:%s:eth0:off",
+		nodeReq.IP,
+		clusterReq.Network.GatewayAddr,
+		net.IP(clusterReq.Network.CIDR.Mask),
+		nodeReq.Name))
+
+	ones, bits := clusterReq.Network.CIDR.IP.DefaultMask().Size()
+
+	cfg := firecracker.Config{
```

Comment:
from this line down I want to extract into separate process which receives `firecracker.Config` serialized and uses it to run the firecracker VM